components/core/src/clp/aws/AwsAuthenticationSigner.cpp

```diff
@@ -0,0 +1,337 @@
+            encoded_uri += c;
+        } else if (c == '/' && false == encode_slash) {
+            encoded_uri += c;
+        } else {
+            encoded_uri += fmt::format("%{:02X}", c);
+        }
+    }
+
+    return encoded_uri;
+}
+
+/**
+ * Gets the scope specified by AWS Signature Version 4 format
+ * @param date_string
+ * @param region
+ * @return The scope as a string
+ */
+[[nodiscard]] auto get_scope(string_view date_string, string_view region) -> string {
+    return fmt::format("{}/{}/{}/{}", date_string, region, cS3Service, cAws4Request);
+}
+
+/**
+ * Gets the canonical request string
+ * @param method HTTP method
+ * @param url S3 Url
+ * @param query_string Query string
+ * @return Canonical request
+ */
+[[nodiscard]] auto get_canonical_request(S3Url const& url, string_view query_string) -> string {
+    return fmt::format(
+            "{}\n{}\n{}\n{}:{}\n\n{}\n{}",
+            get_method_string(AwsAuthenticationSigner::HttpMethod::GET),
+            encode_uri(url.get_path(), false),
+            query_string,
+            cDefaultSignedHeaders,
+            url.get_host(),
+            cDefaultSignedHeaders,
+            cUnsignedPayload
+    );
+}
+}  // namespace
+
+S3Url::S3Url(string const& url) {
+    // Virtual-hosted-style HTTP URL format
+    std::regex const host_style_url_regex(
+            R"(https://([a-z0-9.-]+)\.s3(\.([a-z0-9-]+))?\.amazonaws\.com(/[^?]+).*)"
+    );
+    // Path-style HTTP URL format
+    std::regex const path_style_url_regex(
+            R"(https://s3(\.([a-z0-9-]+))?\.amazonaws\.com/([a-z0-9.-]+)(/[^?]+).*)"
+    );
+
+    std::smatch match;
+    if (std::regex_match(url, match, host_style_url_regex)) {
+        m_bucket = match[1].str();
+        m_region = match[3].str();
+        m_path = match[4].str();
+    } else if (std::regex_match(url, match, path_style_url_regex)) {
+        m_region = match[2].str();
+        m_bucket = match[3].str();
+        m_path = match[4].str();
+    } else {
+        throw OperationFailed(
+                ErrorCode_BadParam,
+                __FILENAME__,
+                __LINE__,
+                "Invalid S3 HTTP URL format: " + url
+        );
+    }
+
+    if (m_region.empty()) {
+        m_region = cDefaultRegion;
+    }
+    m_host = fmt::format("{}.s3.{}.amazonaws.com", m_bucket, m_region);
+}
+
+auto AwsAuthenticationSigner::generate_presigned_url(
+        S3Url const& s3_url,
+        string& presigned_url
+) const -> ErrorCode {
+    auto const s3_region = s3_url.get_region();
+
+    auto const now = std::chrono::system_clock::now();
+    auto const date_string = get_date_string(now);
+    auto const timestamp_string = get_timestamp_string(now);
+
+    auto const scope = get_scope(date_string, s3_region);
+    auto const canonical_query_string = generate_canonical_query_string(scope, timestamp_string);
+
+    auto const canonical_request = get_canonical_request(s3_url, canonical_query_string);
+
+    string string_to_sign{};
```

Comment:
```suggestion
    string string_to_sign;
```
We should rely on the default constructor if the class has any